flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/FlinkJavaConstantHiveVarcharObjectInspector.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions.hive.conversion;
+
+import org.apache.hadoop.hive.common.type.HiveVarchar;
+import org.apache.hadoop.hive.serde2.io.HiveVarcharWritable;
+import org.apache.hadoop.hive.serde2.objectinspector.ConstantObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.JavaHiveVarcharObjectInspector;
+import org.apache.hadoop.hive.serde2.typeinfo.BaseCharUtils;
+import org.apache.hadoop.hive.serde2.typeinfo.VarcharTypeInfo;
+
+/**
+ * Our own implementation of JavaConstantHiveVarcharObjectInspector that takes a TypeInfo in constructor, and overrides
+ * getPrimitiveJavaObject.
+ */
+public class FlinkJavaConstantHiveVarcharObjectInspector
+		extends JavaHiveVarcharObjectInspector implements ConstantObjectInspector {
+
+	private final HiveVarchar value;
+
+	public FlinkJavaConstantHiveVarcharObjectInspector(VarcharTypeInfo varcharTypeInfo, HiveVarchar value) {
+		super(varcharTypeInfo);
+		this.value = value == null || BaseCharUtils.doesPrimitiveMatchTypeParams(value, varcharTypeInfo) ?
```

Comment:
BaseCharUtils.doesPrimitiveMatchTypeParams() seems check if the value length is equal to the precision, which is valid for CHAR type, but not for VARCHAR.

Also, I saw inconsistency between this one and doesWritableMatchTypeParams().

Can we doubt check on this?

Reply:
You're right. We only need to enforce the length if character length exceeds the specified max length.